internal/state/manager.go

```diff
@@ -0,0 +1,138 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT
+
+package state
+
+import (
+	"errors"
+	"log"
+	"os"
+	"time"
+)
+
+const (
+	// defaultSaveInterval is the default duration between state file saves
+	defaultSaveInterval = 100 * time.Millisecond
+	// defaultQueueSize is the default capacity of the offset queue
+	defaultQueueSize = 2000
+)
+
+type rangeManager struct {
+	name            string
+	stateFilePath   string
+	queue           chan Range
+	saveInterval    time.Duration
+	maxPersistItems int
+	replaceTreeCh   chan *rangeTree
+}
+
+// FileRangeManager is a state manager that handles the Range.
+type FileRangeManager Manager[Range, RangeList]
+
+var _ FileRangeManager = (*rangeManager)(nil)
+
+func NewFileRangeManager(cfg ManagerConfig) FileRangeManager {
+	if cfg.QueueSize <= 0 {
+		cfg.QueueSize = defaultQueueSize
+	}
+	if cfg.SaveInterval <= 0 {
+		cfg.SaveInterval = defaultSaveInterval
+	}
+	return &rangeManager{
+		name:            cfg.Name,
+		stateFilePath:   cfg.StateFilePath(),
+		queue:           make(chan Range, cfg.QueueSize),
+		saveInterval:    cfg.SaveInterval,
+		maxPersistItems: cfg.MaxPersistItems,
```

Comment:
For the case where `maxPersistItems` is 1, I'm thinking about changing the logic so it doesn't actually insert it into the tree and just keeps track of the maxOffset. Should be more performant.